src/elastio-snap.c

```diff
@@ -1830,29 +1833,35 @@ int file_write_block(struct snap_device *dev, void *block, size_t offset, size_t
 	struct block_device *bdev;
 	sector_t start_sect;
 	int sectors_processed;
+	int iterations_done;
+	int bytes_written;
 
 	ret = 0;
 	bs = dev_bioset(dev);
 	bdev = dev->sd_base_dev;
 	sectors_processed = 0;
 
-	WARN_ON(!IS_ALIGNED(offset, PAGE_SIZE) || len > SECTORS_PER_BLOCK);
-
-	if (dev->sd_cow)
-		file_unlock(dev->sd_cow->filp);
+	WARN_ON(len > SECTORS_PER_BLOCK);
 
 write_bio:
 	start_sect = sector_by_offset(dev, offset);
+	if (start_sect == SECTOR_INVALID) {
+		LOG_WARN("Possible write IO to the end of file (offset=%lu)", offset);
+		goto out;
```

Comment:
No need to return errors, all critical events are handled by outside logic. Despite that, there is a valid scenario, when this return is needed. Suppose you're trying to write the last block (4k, 8 sectors) to the disk, but there are only a couple of sectors left. If you don't handle this, you will end up with 'attempt write beyond the end of device', which is definitely not something you'd like to see.